Holoscan Flow Benchmarking Logging Updates

tbirdso opened the pull request:
Changes:
- Update Holoscan Flow Benchmarking `benchmark.py` to use standard Python [logging](https://docs.python.org/3/library/logging.html) module
- Increase missing log file severity from warning to error
- Fix paths for log file existence check to consider specified log directory

Log files may be missing if the underlying `./run launch` command fails to run the application, such as in the event of a bad app configuration or run environment.

sohamm17 commented on benchmarks/holoscan_flow_benchmarking/benchmark.py:
```diff
@@ -272,19 +275,24 @@ def main():
+            logging.info(f"Run {i} completed for {scheduler} scheduler.")
             time.sleep(1)  # cool down period
 
+    def print_missing(files: list) -> None:
+        absolute_filepaths = [os.path.abspath(os.path.join(log_directory, file)) for file in files]
+        missing_files = [file for file in absolute_filepaths if not os.path.exists(file)]
+        if missing_files:
+            logging.error(f"Log file(s) {', '.join(missing_files)} are missing.")
```
```suggestion
    def print_missing(files: list) -> bool:
        absolute_filepaths = [os.path.abspath(os.path.join(log_directory, file)) for file in files]
        missing_files = [file for file in absolute_filepaths if not os.path.exists(file)]
        if missing_files:
            logging.error(f"Log file(s) {', '.join(missing_files)} are missing.")
            return true
        return false  
```

Something like the above could be done.

tbirdso replied:
Thanks for the suggestion, I agree with the goal here, it is useful to have missing/present file results available for more granular consideration. Removed this function in favor of an inlined block.

sohamm17 reviewed (commented):
@tbirdso Great improvements! I have left a few suggestions which are necessary, in my opinion.

Also, could you kindly show some sample outputs of the `logging` module in python that we are getting it here.

tbirdso commented on benchmarks/holoscan_flow_benchmarking/benchmark.py:
```diff
@@ -272,19 +275,24 @@ def main():
+    logging.info("\nEvaluation completed.")
+    logging.info(f"Log file directory: {os.path.abspath(log_directory)}")
+    logging.info(f"All the data flow tracking log files are: {', '.join(log_files)}")
+    print_missing(log_files)
```
Agreed, it is a better dev experience to print each file once as missing or not missing. Updated accordingly.

tbirdso commented on benchmarks/holoscan_flow_benchmarking/benchmark.py:
```diff
@@ -272,19 +275,24 @@ def main():
-        for missing_file in (file for file in gpu_utilization_log_files):
-            print(f"Warning: Utilization log file {missing_file} is missing.")
+        logging.info("All the GPU utilization log files are:", ", ".join(gpu_utilization_log_files))
+        print_missing(gpu_utilization_log_files)
```
Updated

tbirdso commented on benchmarks/holoscan_flow_benchmarking/benchmark.py:
```diff
@@ -272,19 +275,24 @@ def main():
+        absolute_filepaths = [os.path.abspath(os.path.join(log_directory, file)) for file in files]
+        missing_files = [file for file in absolute_filepaths if not os.path.exists(file)]
+        if missing_files:
+            logging.error(f"Log file(s) {', '.join(missing_files)} are missing.")
```
Agreed with the goal here, it's useful to control script behavior based on whether files are missing. Refactored and updated to exit with error if files are missing at the end of execution.

tbirdso commented:
Added quality-of-life updates:
- Cleaned up output file existence log messages so that it is clear whether files are present or missing, and updated script behavior to exit with error if output files are missing
- Added `--level` CLI argument to control console output verbosity
- Added file handler with `logging.DEBUG` severity to capture verbose output in the specified log directory (`benchmark.log`)
- Log `stdout` and `stderr` from each returned concurrent instance with `DEBUG` verbosity to aid log file failure investigations
- Updated HoloHub `run` script to propagate exit code from application command so that benchmarking can detect process failures (@jjomier please review)

Failed run example (headless SSH):
```
$ python ./benchmarks/holoscan_flow_benchmarking/benchmark.py -a endoscopy_tool_tracking  --sched greedy -i2
2024-11-08 06:34:25,079 INFO: Run 1 started for greedy scheduler.
2024-11-08 06:34:25,375 ERROR: Command "./run launch endoscopy_tool_tracking cpp" exited with code 134
2024-11-08 06:34:25,398 ERROR: Command "./run launch endoscopy_tool_tracking cpp" exited with code 134
2024-11-08 06:34:25,398 INFO: Run 1 completed for greedy scheduler.
2024-11-08 06:34:26,398 INFO: ****************************************************************
2024-11-08 06:34:26,398 INFO: Evaluation completed.
2024-11-08 06:34:26,398 INFO: ****************************************************************
2024-11-08 06:34:26,399 INFO: Log file directory: /workspace/holohub/log_directory_20241108-063425
2024-11-08 06:34:26,399 INFO: Log files are available: 
2024-11-08 06:34:26,399 ERROR: Log files are missing: /workspace/holohub/log_directory_20241108-063425/logger_greedy_1_1.log, /workspace/holohub/log_directory_20241108-063425/logger_greedy_1_2.log
2024-11-08 06:34:26,399 ERROR: Some log files are missing. Please check the log directory.
```

Corresponding `benchmark.log` with `stdout`, `stderr` output:
[benchmark.20241108-063425.log](https://github.com/user-attachments/files/17674367/benchmark.20241108-063425.log)

Successful run example, with an additional run and GPU utilization enabled for purpose of demonstration:
```
$ xvfb-run -a python ./benchmarks/holoscan_flow_benchmarking/benchmark.py -a endoscopy_tool_tracking  --sched greedy -i2 -r2 -u
2024-11-08 06:35:18,303 INFO: Run 1 started for greedy scheduler.
2024-11-08 06:35:18,304 INFO: Monitoring GPU utilization in a separate thread
2024-11-08 06:35:22,322 INFO: Run 1 completed for greedy scheduler.
2024-11-08 06:35:23,323 INFO: Run 2 started for greedy scheduler.
2024-11-08 06:35:23,324 INFO: Monitoring GPU utilization in a separate thread
2024-11-08 06:35:27,333 INFO: Run 2 completed for greedy scheduler.
2024-11-08 06:35:28,334 INFO: ****************************************************************
2024-11-08 06:35:28,335 INFO: Evaluation completed.
2024-11-08 06:35:28,335 INFO: ****************************************************************
2024-11-08 06:35:28,335 INFO: Log file directory: /workspace/holohub/log_directory_20241108-063518
2024-11-08 06:35:28,335 INFO: Log files are available: /workspace/holohub/log_directory_20241108-063518/logger_greedy_1_1.log, /workspace/holohub/log_directory_20241108-063518/logger_greedy_1_2.log, /workspace/holohub/log_directory_20241108-063518/logger_greedy_2_1.log, /workspace/holohub/log_directory_20241108-063518/logger_greedy_2_2.log
2024-11-08 06:35:28,335 INFO: Gpu files are available: /workspace/holohub/log_directory_20241108-063518/gpu_utilization_greedy_1.csv, /workspace/holohub/log_directory_20241108-063518/gpu_utilization_greedy_2.csv
```

Corresponding `benchmark.log` with `stdout`, `stderr` output: 
[benchmark.20241108-063518.log](https://github.com/user-attachments/files/17674370/benchmark.20241108-063518.log)

Failed run with console logging limited to errors only:
```
$ python ./benchmarks/holoscan_flow_benchmarking/benchmark.py -a endoscopy_tool_tracking  --sched greedy --level error
2024-11-08 06:40:55,442 ERROR: Command "./run launch endoscopy_tool_tracking cpp" exited with code 134
2024-11-08 06:40:56,452 ERROR: Log files are missing: /workspace/holohub/log_directory_20241108-064055/logger_greedy_1_1.log
2024-11-08 06:40:56,452 ERROR: Some log files are missing. Please check the log directory.
```

These updates aim to improve the `benchmark.py` CLI experience and make it easier to trace flow tracking runtime errors across runs.

sohamm17 reviewed (approved):
LGTM
